### ISSUE TITLE
Bookmarks - with Permalinks?

### ISSUE THREAD

**Issue by jukos** [andreaordonselli/qgis2o.gis#19]
Hello,
is it possible to use permalinks for the bookmark function?
Possible advantages:
- quick creation
- Use of all the advantages of permalinks

Best regards

**Comment by jukos** [comment]
Hello.
I wanted to try something myself, but at the moment I can't activate the bookmark function. I have changed
`<!-- bookmark`
in
`<!-- bookmark -->`

Unfortunately, the bookmark symbol does not appear.
Has anything changed with version 2.23?

Best regards

**Comment by waseemahmed345** [comment]
here is a link of [Bookmark](https://www.opengis.it/docs/o-gis/o-gis-configura-la-mappa/#8-toc-title) tutorial!

**Comment by jukos** [comment]
Yes I know. Can you confirm that it works in version 2.23? If so, with which Joomla version?

**Comment by andreaordonselli** [comment]
Hi, nothing has changed in version 2.23 and the version of your CMS (joomla, wordpress....) does not affect the project.
After uncommenting the bookmarks text the icon should appear in your map and then you need to configure your bookmarks list in `bm.setBookmarks({`

I will let you know if it is possible to write a permalink linked to an already configured bookmark, so that you can zoom directly to the area of interest. Regards.

**Comment by jukos** [comment]
I have found out why the icon was not displayed. If the sample contents under bm.setBookmarks are not deleted, the bookmark icon is not displayed. In the past, however, the icon was displayed. Tested it myself.
However, this was a side issue to the actual question of whether it is possible to use permalinks for the bookmark function. Let's see if it is possible.

**Comment by jukos** [comment]
I have found out why the icon was not displayed. If the sample contents under bm.setBookmarks are not deleted, the bookmark icon is not displayed. In the past, however, the icon was displayed. Tested it myself.
However, this was a side issue to the actual question of whether it is possible to use permalinks for the bookmark function. Let's see if it is possible.

**Comment by andreaordonselli** [comment]
This probably happens because the default example code shows that you can provide coordinates in EPSG (3004) and convert them to the project coordinates (EPSG 25832 in the example). But these EPSGs do not match your project, which is why the icon does not show up.

**Comment by jukos** [comment]
That's right. In my case it works if I enter 3857 for both values.

**Comment by andreaordonselli** [comment]
In the next release this functionality will be integrated, for now edit the index.html taking inspiration from this code

```javascript
<!-- bookmark -->	
<script> 
		bm.setBookmarks({
			Quartiere_1: {pos:ol.proj.transform([2351142,4842857], 'EPSG:3004', 'EPSG:25832'), zoom:15},
			Quartiere_2: {pos:ol.proj.transform([2352404,4841605], 'EPSG:3004', 'EPSG:25832'), zoom:15},
			Quartiere_3: {pos:ol.proj.transform([2355410,4838774], 'EPSG:3004', 'EPSG:25832'), zoom:15},
			Quartiere_4: {pos:ol.proj.transform([2357220,4839534], 'EPSG:3004', 'EPSG:25832'), zoom:15},
			});
		map.addControl(bm);	  
		  
		//Permalink query bookmark
		if (permalink.hasUrlParam('book')) {				
			var urlBookParam = permalink.getUrlParam('book');					
			var bookmarkElement = $(".ol-bookmark ul li[data-name='" + urlBookParam + "']");
			bookmarkElement.trigger('click');
			// Remove "book" parameter from URL using setUrlParam without specifying value
			permalink.setUrlParam('book');	
		}			  
</script>
```

To properly write a permalink you need to:
- capture the link to your project
- delete the data written after the zoom definition
- add `&book=BookmarkName`

You can test these links:

ORIGINAL LINK:
https://www.opengis.it/ente/qgis2o.gis/index.html?lon=12.956319&lat=43.705298&z=14.2

LINK WITH BOOKMARK SEARCH (Quartiere_1) :
https://www.opengis.it/ente/qgis2o.gis/index.html?lon=12.956319&lat=43.705298&z=14.2&book=Quartiere_1

**Comment by jukos** [comment]
I'm sure your code is good. But I can't get it to work....
Would you mind doing this:

Write the full code for bookmark with only one permalink:
https://www.opengis.it/ente/qgis2o.gis/index.html?lon=12.956319&lat=43.705298&z=14.2&book=Quartiere_1

Maybe then I'll understand it.

**Comment by andreaordonselli** [comment]
write me your code so I can indicate what to

**Comment by jukos** [comment]
It may well be that I have not expressed myself correctly. My idea was this:

```
<!-- bookmark -->	
<script> 
var bm = new ol.control.GeoBookmark({
			namespace: 'demo',
			title: "Zoom Localita'",
			editable: false,
		  });
		bm.setBookmarks({
			
			//Instead of this:
			Test: {pos:ol.proj.transform([949063.22, 6354777.68], 'EPSG:3857', 'EPSG:3857'), zoom:17},
			
			//That:			
						
			//Test: https://www.website.de/markt/?lon=8.528262&lat=49.468465&z=15.9&l=basemapdeWebRasterGrau:1|TopPlusOpenLightGrau:1|Farbabdeckung:1|Farbfilter:1|Gebude1StockGebude1:1|Gebude:1|Gebude_Neu:1|Gebude_Neueingefgt:1|Gebude_Neu2gebude_neu__eingefgtHiermchteichmaleinenlngerenTitelausprobieren:1|Gebudeschatten:1|groeZelteZentroide:1|groeZelte:1|groeZelte2groe_zelte_kopie:1|Nummern:1|Zelte_Neuzelte_neu_oben:1|Zelte_Neu_oben:1|ZentroidPunkteausZelte_Neu_oben:1|Veranstaltung:1|Areale:1
			
						
			//Or, if not possible, that:		
			//Test: https://www.website.de/markt/?lon=8.528262&lat=49.468465&z=15.9
			
			
			
			
			//My idea was this: (Link generated with the permalink function)
			
		
			
			});
		map.addControl(bm);	  
		  
		 
</script>
```

Only a possibility to create bookmarks faster, than to pick coordinates.

**Comment by andreaordonselli** [comment]
Define your bookmarks (in example the name is "Test"):

```javascript
<script> 
var bm = new ol.control.GeoBookmark({
			namespace: 'demo',
			title: "Zoom Localita'",
			editable: false,
		  });
		bm.setBookmarks({
			Test: {pos:ol.proj.transform([949063.22, 6354777.68], 'EPSG:3857', 'EPSG:3857'), zoom:17},
			});
		map.addControl(bm);
		
	       //Permalink query bookmark
		if (permalink.hasUrlParam('book')) {				
			var urlBookParam = permalink.getUrlParam('book');					
			var bookmarkElement = $(".ol-bookmark ul li[data-name='" + urlBookParam + "']");
			bookmarkElement.trigger('click');
			// Remove "book" parameter from URL using setUrlParam without specifying value
			permalink.setUrlParam('book');	
		}
</script>
```

Then your link will be this (with &book=Test after zoom difining)
https://www.website.de/markt/?lon=8.528262&lat=49.468465&z=15.9&book=Test

**Comment by jukos** [comment]
If I understand this correctly, the coordinates are still required to create a bookmark.
Furthermore, it is possible to produce the shorter link.
Do I understand that correctly?
Best regards

**Comment by andreaordonselli** [comment]
The permalink you set must have at least coordinates and zooms, and basically set your book=....
This permalink at startup launches the bookmark you have chosen, so as soon as you start the system it will change the coordinates and zoom as you have set them in the bookmark.
Let me know if you have any other doubts.

**Comment by jukos** [comment]
I certainly have no doubts about your work.
And you can certainly leave it as you have programmed it.
I suspect that my idea is not feasible. I was just thinking about whether the bookmark function could be simplified. 
Example:
Instead of
`{pos:ol.proj.transform([949063.22, 6354777.68], 'EPSG:3857', 'EPSG:3857'), zoom:17}`
should be entered (example):
`https://www.website.de/markt/?lon=8.528262&lat=49.468465&z=15.9&l=basemapdeWebRasterGrau:1|TopPlusOpenLightGray:1|Building_New2building_new__inserted..........`

Because this permalink can be generated very quickly with the permalink function and copied into the code.

It was just the thought that you could generate many bookmarks very quickly. However, there is nothing wrong with the existing function.

Best regards

**Comment by andreaordonselli** [comment]
Now I understand what you meant. Unfortunately no, it is not possible to write bookmarks by entering a permalink instead of the default code.
You must use the method indicated in opengis.it/docs for set bookmarks

**Comment by andreaordonselli** [comment]
This issue has been resolved in the v2.24.0 release
Thanks for your report
